src/GraphQl/Subscription/SubscriptionManager.php

```diff
@@ -0,0 +1,124 @@
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * Manages all the queried subscriptions by creating their ID
+ * and saving to a cache the information needed to publish updated data.
+ *
+ * @experimental
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+final class SubscriptionManager implements SubscriptionManagerInterface
+{
+    use IdentifierTrait;
+    use ResourceClassInfoTrait;
+    use SortTrait;
+
+    private $subscriptionsCache;
+    private $subscriptionIdentifierGenerator;
+    private $serializeStage;
+    private $iriConverter;
+
+    public function __construct(CacheItemPoolInterface $subscriptionsCache, SubscriptionIdentifierGeneratorInterface $subscriptionIdentifierGenerator, SerializeStageInterface $serializeStage, IriConverterInterface $iriConverter)
+    {
+        $this->subscriptionsCache = $subscriptionsCache;
+        $this->subscriptionIdentifierGenerator = $subscriptionIdentifierGenerator;
+        $this->serializeStage = $serializeStage;
+        $this->iriConverter = $iriConverter;
+    }
+
+    public function retrieveSubscriptionId(array $context, ?array $result): ?string
+    {
+        /** @var ResolveInfo $info */
+        $info = $context['info'];
+        $fields = $info->getFieldSelection(PHP_INT_MAX);
+        $this->arrayRecursiveSort($fields, 'ksort');
+        $iri = $this->getIdentifierFromContext($context);
+        if (null === $iri) {
+            return null;
+        }
+        $subscriptionsCacheItem = $this->subscriptionsCache->getItem($this->encodeIriToCacheKey($iri));
+        $subscriptions = [];
+        if ($subscriptionsCacheItem->isHit()) {
+            $subscriptions = $subscriptionsCacheItem->get();
+            foreach ($subscriptions as [$subscriptionId, $subscriptionFields, $subscriptionResult]) {
+                if ($subscriptionFields === $fields) {
+                    return $subscriptionId;
+                }
+            }
+        }
+
+        $subscriptionId = $this->subscriptionIdentifierGenerator->generateSubscriptionIdentifier();
+        unset($result['clientSubscriptionId']);
+        $subscriptions[] = [$subscriptionId, $fields, $result];
+        $subscriptionsCacheItem->set($subscriptions);
+        $this->subscriptionsCache->save($subscriptionsCacheItem);
+
+        return $subscriptionId;
+    }
+
+    /**
+     * @param object $object
+     */
+    public function getPushPayloads($object): array
+    {
+        $iri = $this->iriConverter->getIriFromItem($object);
+        $subscriptions = $this->getSubscriptionsFromIri($iri);
+
+        $resourceClass = $this->getObjectClass($object);
+
+        $payloads = [];
+        foreach ($subscriptions as [$subscriptionId, $subscriptionFields, $subscriptionResult]) {
+            $resolverContext = ['fields' => $subscriptionFields, 'is_collection' => false, 'is_mutation' => false, 'is_subscription' => true];
+
+            $data = ($this->serializeStage)($object, $resourceClass, 'update', $resolverContext);
```

Comment:
@alanpoulain This means that the subscription data will get serialized using the `update` context, right? Should there be a way to override it for the subscription payload specifically?

Reply:
I think it would be a bad idea to have a `subscription` context. Subscriptions are here to send the updated data like you would receive it if you made a mutation and read the received data.

Reply:
You might be right 😉  what about the case where there is no `update` mutation on the resource? (It can still be changed from other parts of the API, right?)

Reply:
Would you see the `item_query` context instead then?

Reply:
@alanpoulain Sorry for the late reply - I think trying `update` first and falling back to `item_query` would be good, wdyt?

Reply:
Actually, if you don't have an `update` operation, the subscription fields are not generated, so it makes sense to me to keep it like this.